### PR TITLE
WIP Set up project permissions for user

### PR DESCRIPTION
**Pull request by amyd11**
Story: https://jiraent.cms.gov/browse/ISPGBSS-863

**_WIP will be removed when print statements come out, but they are left in for now while testing and clarity for devs reviewing this_**
______
### The Lowdown
_These changes will set up individual project permissions for users. A set of Groups will manage the permissions on a project and the user may be added to a Group for each project that they need to access and thus inherit the permissions based on their Group(s)._

- Groups will be created dynamically once a project is initially created.
- These groups will be named based on the project id (e.g. 10_project_admin_group).
- There are three groups for each project (project admin, project collaborator, project viewer)
- Each group will have specific permissions as follows

|   | Project Admin | Project Contributor | Project Viewer |
|--|---------------|--------------------|-----------------|
| Has read access | X | X | X |
| Has write access | X | X |   |
| Has admin access | X |   |   |
- These permissions will be object level permissions, so will pertain to each project individually. (Django Guardian has been used to set up and manage these object level permissions for the groups)
- The user that originally created the project will be added to the admin group for that project initially
- This work sets up the permission ability for users to be added or deleted from groups for multiple projects as more user management around projects is developed

### Changes
- set up a permission_constants.py file to house the constants related to creating names for groups & permissions and mappings
- added a utils file to include a function that will use the permission constants to generate group names and assign object level permissions to those groups (by using django guardian)
- update the Package model to include a `create_groups_for_project` function which will create groups (project admin, project contributor, project viewer) for the project instance then will assign the created project admin group to the user
- updated project/views.py to query for the set of projects that the user has access to based on their permissions. This query uses Django Guardian since the permissions are through the user's groups and are object level permissions managed via Guardian.

### Testing
_Try testing this against a user that is not a superuser. If you want to test this against a user that is a superuser, you will need to make the following modification in code, because otherwise any superuser has access to all projects._
> Go to the `project/views.py` file > `class PackageList` > def get_queryset
then make the following change in the query for generating the `projects_list`
add the parameter: `with_superuser=False`
- First, go to /packages to compare what projects you currently have access to
- Then, go to /packages/create and fill out the form to create a project
- Go back to /packages to see the new project show up

You can check groups and permissions for the new project by looking into your local database:
- First lets check the project id that we just created. Look at the `project_package` table for the project that was created. You'll see something like:

> | table entry id | project id | title | creator (user id of the creator) |
> |-----|-----|-----|---|
> | 33 | **86** | Excited Elephant | **2** |

- Look at the `auth_group` table which shows the list of groups that users belong to. You should have 3 new entries all which have group names prefaced by the new project id. e.g.

> | table entry id | group id | group name |
> |-----|-----|-----|
> | 19 | **43** | 86_project_admin_group |
> | 20 | **44** | 86_project_contributor_group |
> | 21 | **45** | 86_project_view_only_group |

- You can cross check the group id from the above table with your logged in user that just created the new project. Look at the `siteapp_user_groups` table. You should see your user id match up with the group id for the project admin group. Meaning that your user is a member of that project admin group. e.g.

> | table entry id | user id | group id |
> |-----|-----|-----|
> | 6	| **2**	|**43**|

- Next lets check the permissions on the groups that were just created. Since these groups have object level permissions on a project, django guardian handles that so we're going to look at the `guardian_groupobjectpermission` table

> | table entry id | object_pk (project id) | content type id | group id | permission id |
> | -- | -- | -- | -- | -- |
> |	34| 	86	| 61	| 43	| 273| 
> |	35| 	86	| 61	| 43	| 274| 
> |	36| 	86	| 61	| 43	| 275| 
> |	37| 	86	| 61	| 44	| 274| 
> |	38| 	86	| 61	| 44	| 275| 
> |	39| 	86	| 61	| 45	| 275| 

_So we have 6 new entries total. Group 43 (the project admin group) has 3 new entries for permission ids 273, 274, 275. Group 44 (the project contributor group) has 2 new entries. And group 45 (the project viewer group) has 1 new entry._

- We can see what permissions these map to by looking in the `auth_permission` table

> | table entry id | permission id | name (permission description) | content type id | codename |
> | -- | -- | -- | -- | -- |
> | 258 | 273 | Can manage users on project | 61 | project_can_manage_users |
> | 259 | 274 | Can edit project | 61 | project_can_edit |
> | 260 | 275 | Can view project | 61 | project_can_view |

That's it!

### CODE REVIEW

File: project/views.py [hunk 1]
```diff
@@ -43,20 +59,6 @@ def project(request, project_id):
     )
 
 
-# TODO Example def that is not yet functional
-@login_required
-def project_component_detail(request, project_id, component_id):
```
Comment (TMCamp) [inline project/views.py:48]: The URL for this still exists in the urls.py line 12 so it caused an error when I ran `run.py`.

File: project/models.py [hunk 2]
```diff
@@ -66,16 +78,53 @@ class Package(models.Model):
...
+def create_groups_for_project(sender, instance, **kwargs):
+    print("*----- create_groups_for_project function -----*")
+    pprint("---sender")
+    pprint(sender)
+    pprint("---instance")
+    pprint(instance)
+    pprint("---kwargs")
+    pprint(kwargs)
+
+    # only want to do this when a project is created
```
Comment (TMCamp) [inline project/models.py:105]: Is `created` a boolean value that always exists in the kwargs? If not you may want to use `if "created" in kwargs:`

**Review by TMCamp — CHANGES_REQUESTED** [review]

File: project/models.py (hunk 2 above)
Comment (amyd11) [reply project/models.py:105]: Ah I like that suggestion, but it does show up as a key with a boolean value. Here's the docs for the post_save https://docs.djangoproject.com/en/3.2/ref/signals/#post-save

This is an example of what we get:
```
{'created': True,
 'raw': False,
 'signal': <django.db.models.signals.ModelSignal object at 0x7fdead30f430>,
 'update_fields': None,
 'using': 'default'}
```

File: project/views.py (hunk 1 above)
Comment (amyd11) [reply project/views.py:48]: Thanks for catching that... I actually just took all the placeholder urls out for now.

**Pull request closed without merging**